R/credentials.R

```diff
@@ -0,0 +1,365 @@
+
+  creds_tbl <- get_keyring_creds_table()
+
+  creds_tbl <-
+    creds_tbl %>%
+    dplyr::filter(
+      key_name == key_name,
+      version == keyring_schema_version
+    )
+
+  if (nrow(creds_tbl) == 0) {
+
+    stop("There are no keyring credentials available for the `key_name` of \"",
+         key_name, "\".",
+         call. = FALSE)
+  }
+
+  service_name <- creds_tbl %>% dplyr::pull(service)
+
+  keyring::key_get(service = service_name) %>%
+    jsonlite::unserializeJSON()
+}
+
+#' Helpers for supplying SMTP credentials
+#'
+#' These helper functions, the credential helpers, are used to supply SMTP
+#' configuration and authorization information for the [smtp_send()] function.
+#' These helpers ([creds_file()], [creds_key()], and [creds()]) are to be
+#' used expressly with the `credentials` argument of [smtp_send()]. The
+#' [creds_file()] credential helper is used to point at a credentials file
+#' stored on disk. We can create that file using the [create_smtp_creds_file()]
+#' function. The [creds_key()] credential helper is truly somthing special.
+#' It actually helps one obtain credentials that are stored in the system-wide
+#' key-value store. The equally magical [create_smtp_creds_key()] function sets
+#' that key. Using [creds()] allows for manual specification of SMTP
+#' configuration and authentication within that helper function.
+#'
+#' @param file When using the [creds_file()] credential helper, we need to
+#'   specify the location of the credential file, and, this is where that is
+#'   done. The credential file was ideally generated by the
+#'   [create_smtp_creds_file()] function.
+#' @param key_name When using the [creds_key()] credential helper, the name of
+#'   the key (in the system key-value store) needs to be given here. This was
+#'   either explicitly provided when using the [create_smtp_creds_key()]
+#'   function (with its own `key_name` argument), or, it may have been
+#'   automatically generated based on the SMTP host name.
+#' @inheritParams create_smtp_creds_file
+#' @name credential_helpers
+#' @return a list object of class `location_cells`.
+NULL
+
+#' Credentials helper for pointing to a SMTP credentials file
+#'
+#' @rdname credential_helpers
+#' @export
+creds_file <- function(file) {
+
+  structure(
+    class = c("creds_file", "blastula_creds"),
+    list(
+      file = file
+    )
+  )
+}
+
+#' Credentials helper for pointing to a SMTP credentials key
+#'
+#' @rdname credential_helpers
+#' @export
+creds_key <- function(key_name) {
+
+  structure(
+    class = c("creds_key", "blastula_creds"),
+    list(
+      key_name = key_name
+    )
+  )
+}
+
+#' Credentials helper for manual entry of SMTP config and auth values
+#'
+#' @rdname credential_helpers
+#' @export
+creds <- function(provider = NULL,
+                  user,
+                  sender_name = NULL,
+                  host = NULL,
+                  port = NULL,
+                  use_ssl = TRUE,
+                  ...,
```

Comment:
What's the `...` for?

Reply:
This does nothing but force the user to name the subsequent arguments. Barret suggested this and I believe this pattern is used twice in this package.

Reply:
I wouldn’t add ... merely for that reason personally; the downside is that incorrect arguments will be silently accepted and also it's just sort of surprising.

Are you sure we want to allow password arguments at all, and not just insist on interactively prompting?

Reply:
I wouldn’t either (this all came about during code pairing earlier this week). Mostly for the surprising part, and I tend to want to avoid `...` wherever possible. 

I’m totally going to defer to your decisions here: I want the credentials part to be best-practices-minded. I will reverse the inclusion of the `password` argument. I don’t think we should support the non-interactive use of this function.

Reply:
Removed all uses of `...` and also removed the `password` arg everywhere.